Userland/Libraries/LibJIT/X86_64/Assembler.h

```diff
@@ -877,18 +877,34 @@ struct X86_64Assembler {
 
     void native_call(
         u64 callee,
-        Vector<Operand> const& preserved_registers = {},
-        Vector<Operand> const& stack_arguments = {})
+        Vector<Operand> const& arguments,
+        Vector<Operand> const& preserved_registers = {})
     {
         for (auto const& reg : preserved_registers.in_reverse())
             push(reg);
 
-        // Preserve 16-byte stack alignment for non-even amount of stack-passed arguments
-        auto needs_aligning = ((stack_arguments.size() + preserved_registers.size()) % 2) == 1;
+        size_t register_arguments;
+        size_t stack_arguments;
+        if (arguments.size() > REG_ARGS.size()) {
+            register_arguments = REG_ARGS.size();
+            stack_arguments = arguments.size() - REG_ARGS.size();
+        } else {
+            register_arguments = arguments.size();
+            stack_arguments = 0;
+        }
+
+        // Preserve 16-byte stack alignment for non-even amount of stack elements
+        auto needs_aligning = ((stack_arguments + preserved_registers.size()) & 1) == 1;
         if (needs_aligning)
-            push(Operand::Imm(0));
-        for (auto const& stack_argument : stack_arguments.in_reverse())
-            push(stack_argument);
+            sub(Operand::Register(Reg::RSP), Operand::Imm(sizeof(u64)));
+
+        for (size_t i = 0; i < register_arguments; i++)
+            mov(Operand::Register(REG_ARGS[i]), arguments[i]);
```

Comment:
This clobbers the argument registers, which leads to hard to detect bugs when one of the arguments is passed at the wrong index.

Reply:
I'm sorry but I fail to see how this is that much worse than the current status quo of just assuming the right values will be in the right registers and if it really is that much easier to debug.
As also previously discussed in discord, `stack_arguments` would be untruthful as it actually would be `7th_argument_and_onwards`, and on other architectures where the ABI specifies more than 6 argument registers (which both Aarch64 and RiscV64 do) you'd have the same issues as you described, except even more obscured. 
I'm also not super familiar with other architectures and whether we'd want to support them but potentially an architecture which defines *less* than 6 arguments registers would be even more problematic with the current approach, though that's just speculation on my end.

Reply:
It is worse than the current status quo, because it's hard to encode the same bug with the old API.
Sure you could do some incorrect movs into ARGn, but that would be visible in the code doing the mov. Now native_call hides that.
Take for example
```c++
mov(Assembler::Operand::Register(ARG1), Assembler::Operand::Imm(5));
mov(Assembler::Operand::Register(ARG2), Assembler::Operand::Register(ARG1)); // pretty obvious
mov(Assembler::Operand::Register(ARG3), Assembler::Operand::Imm(7));
native_call((void*)cxx_foo);
```
vs
```c++
native_call((void*)cxx_foo, {
                                             Assembler::Operand::Register(VM_POINTER),
                                             Assembler::Operand::Imm(5),
                                             Assembler::Operand::Register(ARG1), // not obvious how this turns into Imm(5)
                                             Assembler::Operand::Imm(7),
                                         });
```

Also if I had to speculate, I'm pretty sure we won't add a platform with less than 6 register arguments any time soon. We removed x86-32 support and future platforms are probably all RISC with plenty of registers meaning calling conventions passing in registers.